File: solutions/019.Remove_Nth_Node_From_End_of_List/AC_two_point_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_two_point_n.py
# Create Date: 2015-03-04 19:16:10
# Usage:       AC_two_point_n.py 
# Descripton:  


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        dummy = ListNode(0)
        dummy.next = head
        p, q = dummy, dummy
        
        # first 'q' go n step
        for i in range(n):
            q = q.next

        # q & p
        while q.next:
            p = p.next
            q = q.next

        rec = p.next
        p.next = rec.next
        del rec
        return dummy.next

File: solutions/109.Convert_Sorted_List_to_Binary_Search_Tree/AC_two_point_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_two_point_n.cpp
*  Create Date: 2015-03-09 10:38:49
*  Descripton:  Use two points.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        if (!head)
            return NULL;
        if (!head->next)
            return new TreeNode(head->val);

        // two points
        ListNode *fast = head, *slow = head, *pre = NULL;
        while (slow && fast && fast->next) {
            pre = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        if (pre)
            pre->next = NULL; // warning: this will change the origin list

        TreeNode *root = new TreeNode(slow->val);
        root->left = sortedListToBST(head);
        root->right = sortedListToBST(slow->next);

        return root;
    }
};

int main() {
    ListNode a(0);
    Solution s;
    s.sortedListToBST(&a);
    return 0;
}

